driver/pace/driver/state.py

```diff
@@ -175,12 +179,116 @@ def _overwrite_state_from_restart(
     return state
 
 
+def _overwrite_state_from_fortran_restart(
+    path: str,
+    communicator: pace.util.CubedSphereCommunicator,
+    state: Union[fv3core.DycoreState, pace.physics.PhysicsState, TendencyState],
+    restart_file_prefix: str,
+    is_gpu_backend: bool,
+):
+    """
+    Args:
+        path: path to restart files
+        rank: current rank number
+        state: an empty state
+        restart_file_prefix: file prefix name to read
+
+    Returns:
+        state: new state filled with restart files
+    """
+
+    state_dict = _driver_state_to_dict(state, restart_file_prefix, is_gpu_backend)
+
+    state_dict = pace.util.open_restart(path, communicator, fortran_restart=True)
+
+    state = _dict_state_to_driver_state(state_dict, state, is_gpu_backend)
+
+    return state
+
+
+def _driver_state_to_dict(
+    driver_state: Union[fv3core.DycoreState, pace.physics.PhysicsState, TendencyState],
+    restart_file_prefix: str,
+    is_gpu_backend: bool,
+):
+    """
+    Takes a Pace driver state
+    and returns a dict of state quantities with their Fortran names
+    """
+
+    dict_state = {}
+
+    for _field in fields(type(driver_state)):
+        if "units" in _field.metadata.keys():
+            if is_gpu_backend:
+                if "phy" in restart_file_prefix:
+                    dict_state[_field.name] = driver_state.__dict__[_field.name][:]
+                else:
+                    dict_state[_field.name] = driver_state.__dict__[_field.name].data[:]
+            else:
+                dict_state[_field.name] = driver_state.__dict__[_field.name].data[:]
+
+    return dict_state
+
+
+def _dict_state_to_driver_state(
+    fortran_state: dict,
+    driver_state: Union[fv3core.DycoreState, pace.physics.PhysicsState, TendencyState],
+    is_gpu_backend: bool,
+):
+    """
+    Takes a dict of state quantities with their Fortran names and a driver state
+    and populates the driver state with quantities from the dict.
+    """
+
+    for _field in fields(type(driver_state)):
+        if "units" in _field.metadata.keys():
+            if is_gpu_backend:
```

Comment:
I think you could set `to_type` in an if-elif on is_gpu_backend and then have the code below only once, instead of copy-pasted twice.